Add pasting URLs on selected text as Markdown links

steffen opened the pull request:
This pull request adds pasting URLs on selected text as Markdown links.

In order to test this new functionality:
1. Copy this URL `https://docs.github.com` or any other URL
1. Go to https://paste-markdown-steffen-testing.vercel.app/examples
1. Write something like: `You can find the docs here.`
1. Select `here`
1. Paste URL with <kbd>cmd + v</kbd>

Result: You should now see a Markdown link: `[here](https://docs.github.com)`

koddsson reviewed (commented):
This looks good to me. Can we add tests for some of this functionality? Also adding an example on the demo page.

steffen commented on src/paste-markdown-link.ts:
```diff
@@ -0,0 +1,54 @@
+  return selectedText.length && isURL(text) ? `[${selectedText}](${text})` : text
+}
+
+const URL_RE = /^https?:\/\//i
+
+function isURL(url: string): boolean {
+  return URL_RE.test(url)
+}
```
@koddsson Great point! I was wondering the same. I was following the pattern from https://github.com/github/paste-markdown/blob/32b7ea3f29ae8f256f9d19768387be42678ddf30/src/paste-markdown-image-link.ts#L71-L75, but I agree that removing the constant simplifies the code here.

steffen commented on src/paste-markdown-link.ts:
```diff
@@ -0,0 +1,54 @@
+}
+
+function hasPlainText(transfer: DataTransfer): boolean {
+  return Array.from(transfer.types).indexOf('text/plain') >= 0
```
@koddsson Great catch! Thank you! (On a side note, I got this also from https://github.com/github/paste-markdown/blob/32b7ea3f29ae8f256f9d19768387be42678ddf30/src/paste-markdown-image-link.ts#L63-L65 😊)

steffen commented on src/text.ts:
```diff
@@ -1,8 +1,8 @@
-export function insertText(textarea: HTMLInputElement | HTMLTextAreaElement, text: string): void {
+export function insertText(textarea: HTMLInputElement | HTMLTextAreaElement, text: string, addNewline = true): void {
```
@koddsson Great point! I've changed it to an option in 9be60839de01e375078f9dd9ab27fabb04e4db82.

steffen commented:
@koddsson Thank you for your quick and great suggestions! 🙇‍♂️

I think I've resolved all your inline suggestions as well as the following:

> Can we add tests for some of this functionality?

Do you think this test is sufficient? https://github.com/github/paste-markdown/pull/23/commits/d4c60b15124155ddaff3ceb3d3a623485248fb4c

> Also adding an example on the demo page.

I've added an example on the demo page here: https://github.com/github/paste-markdown/pull/23/commits/460e1b6371e667807d09f3f2fe43afc7abf31459
You can also see it here: https://paste-markdown-steffen-testing.vercel.app/examples

I'm happy and open for any other suggestions. Thank you!

steffen commented:
As discussed in [this PR](https://github.com/steffen/paste-markdown/pull/1), I changed the default export into named exports in order to allow us to load the new feature separately for feature flagging.

/cc @Bestra @koddsson

koddsson commented on src/index.ts:
```diff
@@ -1,21 +1,34 @@
+function subscribe(el: HTMLElement): Subscription {
   installTable(el)
-  installLink(el)
+  installImageLink(el)
```
What do you think about having all the `install*` functions in here so that consumers that use subscribe will get all the features, but then if they want to conditionally apply features they can import only the installers they plan on using?

steffen replied:
@koddsson I've thought about that as well. Can you confirm that [this line](https://github.com/github/github/pull/193520/files#diff-56fd9ba9b1627b3f8c4d982bc5106317f74d72fb0cdad47f025ac98d3df042efR12) is the only place where we consume the library at GitHub?

Since `subscribe` is now a named export, users of the library have to adjust the usage of the library either way, so yeah, I agree with adding the new feature to subscribe right away. We then only need to adjust [this line](https://github.com/github/github/pull/193520/files#diff-56fd9ba9b1627b3f8c4d982bc5106317f74d72fb0cdad47f025ac98d3df042efR12) as well to use the `add` and `remove` hooks as I do in [the lines below](https://github.com/github/github/pull/193520/files#diff-56fd9ba9b1627b3f8c4d982bc5106317f74d72fb0cdad47f025ac98d3df042efR14-R24) so that the new Link feature is not loaded by default.

steffen replied:
@koddsson I just saw [your proposal](https://github.com/github/github/pull/193520#discussion_r715459105) on using the `add` and `remove` hooks for all functions. I missed that earlier. I'll go with that approach. 👍 Thank you! 🙇‍♂️

steffen replied:
@koddsson I have applied your proposal here: https://github.com/github/paste-markdown/pull/23/commits/1c6b0cd957ececaa6d056c2dab554591dd3a3bca Thank you! 🙇‍♂️

koddsson commented on src/paste-markdown-link.ts:
```diff
@@ -0,0 +1,52 @@
+
+  if (selectionStart > 1) {
+    const previousChars = textarea.value.substring(selectionStart - 2, selectionStart)
+    return previousChars === ']('
```
What do you think about checking for HTML links as well? Since markdown generally supports HTML as well, should we check if we're in a `<a>` tag?

steffen replied:
I've thought about some other edge cases as well, for example let's say you have this issue template:

```
Repository: [URL of repository]
```

If you select `[URL of repository]` and paste an URL on top of it, with the new feature it would now convert it to `[[URL of repository]](url)`.

Though, I'm not sure we should account for all edge cases or if users will learn quickly that they need to delete the selected text first and then paste the URL in order to have the URL as is.

Regarding detecting if the selected text is within an `<a>` tag:
Do you think of the use case where a user would have an <a>-link in their Markdown, such as `<a href="https://some-url">Docs</a>`, and where they would want to change the `href` by selecting the `https://some-url` part and replace it with another URL?

koddsson reviewed (approved):
This looks generally good to me. I had a couple of questions and suggestions.

This is changing the API so this will be a breaking change.

steffen commented:
> This is changing the API so this will be a breaking change.

@koddsson I agree. I've raised the breaking of the API [here](https://github.com/steffen/paste-markdown/pull/1#discussion_r712222785).
@Bestra had the great idea to keep the default export of `subscribe` alongside the new named exports. Unfortunately `rollup.js` did not like that, see [here](https://github.com/steffen/paste-markdown/pull/1#discussion_r714521973).

@koddsson Do you think including a breaking change for 0.4.x is reasonable? Not sure what version schema paste-markdown follows, for semver the major version would need to be bumped, e.g. 0.3.x -> 1.0.0.

steffen commented:
@koddsson The CI is green now. (I fixed the running of it [here](https://github.com/github/paste-markdown/pull/25) and I fixed the eslint error [here](https://github.com/github/paste-markdown/pull/23/commits/7840f7dd3cdc7c900d3b288f30319fee30957693).)

Is there anything missing in order to merge this PR and create a new release? Thank you!

steffen commented:
Friendly bump on this PR @github/ui-frameworks-reviewers as I will be off on Friday and I would love to release and deploy these changes this week. Thank you! 🙇‍♂️

srt32 commented:
👋 I love making it easier to write markdown! Could you share some more of the motivation here? Users can currently highlight text, click cmd-k and get an empty markdown link that they can then fill in with cmd-v.

I personally would find it surprising that cmd-v would not overwrite the text I've selected (and IMO be annoying). I also wonder about making cmd-v be the default (I think it's not configurable?) for the library as a whole as opposed to a configuration option.

WDYT?

steffen commented:
> edit - I've since been educated that most of the internet already does this ;). Carry on!

@srt32 😂 Fair points! I will definitely detail my motivation behind this in our internal staff-shipping post. In short, I believe once you see selected text automagically converted to a link for the first time, you don't want to go back anymore. 😄 Also, I made the functionality feature-flaggable so that we can first enable it for individuals only and then staff-ship it and gather feedback. Thank you for looking into this PR! 🙇‍♂️❤️

steffen commented:
👋 What is remaining to get this PR merged? 🙇‍♂️

srt32 reviewed (approved):
I'm cool to merge this. Do we have a plan for the change in versions?

steffen commented:
> I'm cool to merge this.

🙇‍♂️ 

>  Do we have a plan for the change in versions?

@srt32 I think that would be best. Not sure what version scheme is followed, but this change is a **breaking change** as it changes the API.
The reason for that is that we changed the export from a default export to a named export in order to support loading selected features (for feature flagging).

So it's now:
```js
import {subscribe} from '@github/paste-markdown'
```
instead of
```
import subscribe from '@github/paste-markdown'
```

This reminds me that we need to update the README as well, which I will do in a moment.

@Bestra had the idea to use a default export and a named export. Though, `rollup.js` [didn't like it](https://github.com/steffen/paste-markdown/pull/1#discussion_r714521973). We could maybe configure `rollup.js` to allow it I guess, but we then though the easiest is to just change the API.

srt32 commented:
https://github.com/github/paste-markdown/releases/tag/v1.0.2 🎉

steffen commented:
> https://github.com/github/paste-markdown/releases/tag/v1.0.2 🎉

Woohoo, thank you @srt32! 🎉 🙇‍♂️
